### PR TITLE
Implement resource provider for AWS::Lambda::Function

### PR DESCRIPTION
**Pull request by dominikschubert**
## Motivation

Part of the ongoing effort to migrate to `ResourceProvider` based resources.

Normally we'd add the whole service in a single PR but since the lambda function is pretty big in itself, I thought it makes more sense to introduce this separately.

## Changes

- Introduces a new resource provider for `AWS::Lambda::Function`. (Partially based on work from @joe4dev  in https://github.com/localstack/localstack/pull/8406)
- Adds support for cfnresponse module in python functions (used for simple custom resource communication). See: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-lambda-function-code-cfnresponsemodule.html
- Improve support for cfn-response module in NodeJS functions (proper locations for `cfn-response.js` files and optional detection)
- Support for idempotent delete on a function

### CODE REVIEW

File: localstack/services/awslambda/resource_providers/aws_lambda_function.py [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+    OperationStatus,
+    ProgressEvent,
+    ResourceProvider,
+    ResourceRequest,
+)
+from localstack.services.lambda_.lambda_utils import get_handler_file_from_name
+from localstack.utils.archives import is_zip_file
+from localstack.utils.files import mkdir, new_tmp_dir, rm_rf, save_file
+from localstack.utils.strings import is_base64, to_bytes
+from localstack.utils.testutil import create_zip_file
```
Comment (dominikschubert) [inline localstack/services/awslambda/resource_providers/aws_lambda_function.py:20]: it's not very urgent but at some point these need to be inlined or included in `localstack.services.cloudformation.provider_utils`

**Review by pinzon — APPROVED** [review]
LGTM

File: localstack/services/awslambda/resource_providers/aws_lambda_function.py [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+        code["Architectures"] = properties.get("Architectures")
+    return code
+
+
+class LambdaFunctionProvider(ResourceProvider[LambdaFunctionProperties]):
+
+    TYPE = "AWS::Lambda::Function"  # Autogenerated. Don't change
+    SCHEMA = util.get_schema_path(Path(__file__))  # Autogenerated. Don't change
+
+    def create(
```
Comment (dfangl) [inline localstack/services/awslambda/resource_providers/aws_lambda_function.py:281]: Do I understand correctly that this is called repeatedly as long as the resource creation is in progress, which acts as a waiter of sorts?

**Review by dfangl — APPROVED** [review]
LGTM! Just one question for understanding.

**Comment by simonrw** [comment]
We will need to include the `DeadLetterConfig` parameter: https://github.com/localstack/localstack/pull/9773

File: localstack/services/awslambda/resource_providers/aws_lambda_function.py [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+                model,
+                [
+                    "Architectures",
+                    "Description",
+                    "FunctionName",
+                    "Handler",
+                    "ImageConfig",
+                    "PackageType",
+                    "Layers",
+                    "MemorySize",
```
Comment (Morijarti) [inline localstack/services/awslambda/resource_providers/aws_lambda_function.py:342]: After selection timeout and MemorySize need to be converted to int if they aren't already

File: localstack/services/awslambda/resource_providers/aws_lambda_function.py [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+            "Runtime",
+            "Timeout",
+            "TracingConfig",
+            "VpcConfig",
+        ]
+        update_config_props = util.select_attributes(request.desired_state, config_keys)
+        function_name = request.previous_state["FunctionName"]
+        update_config_props["FunctionName"] = function_name
+        if "Timeout" in update_config_props:
+            update_config_props["Timeout"] = int(update_config_props["Timeout"])
```
Comment (Morijarti) [inline localstack/services/awslambda/resource_providers/aws_lambda_function.py:478]: Same conversion should be done for MemorySize

**Review by Morijarti — CHANGES_REQUESTED** [review]

**Review by Morijarti — APPROVED** [review]

**Pull request merged**